### DISCUSSION TITLE
Version 1.9.0 is Available

### DISCUSSION THREAD

**Discussion by mobsense** [sensedeep/dynamodb-onetable#147] [Show and tell]
Version 1.9.0 is now available.

Read the change log at:

https://github.com/sensedeep/dynamodb-onetable/releases/tag/v1.9.0

This release fixes all outstanding issues and adds support for persisting schemas in the table to setup for better tooling.

## Breaking Compatibility

There are a few breaking changes we've been holding off on for a while. These are mostly minor and will impact users that use the `transform` schema property, and those who use the _type field by default in returned items.

The breaking changes are:

* The type field is now hidden by default. Use {hidden: true} in the params to return it. If using fetch or groupByType, you will probably need to include {hidden: true} in the params to return items with the type.
* Transforms are refactored to deprecate Table.intercept and rename as Table.transform, remove Params.transform (was undocumented) and remove Schema Fields.transform and add Params.transform in API calls.

**Comment by rjmackay** [comment]
We're using transform, primarily to change Luxon DateTime objects into strings. Is it possible to add custom field types? That would actually be a much more elegant solution that what we current do, and seems like it would work better with a serializable schema.

**Reply by mobsense** [reply]
We've converted the internals to use "string" types instead of function constructors, so that is possible.

The issues are that the set of types being passed to and from OneTable is the set of JavaScript types. The "type" schema field really describes a mapping from a JavaScript type to DynamoDB. The transform function is used to provide a custom formatting, but the user-code type is still one of the standard JavaScript types and the storage is of course in a DynamoDB supported type.

I'd have to dig and see what places are relying on full understanding the JavaScript type. 

Can you propose a bit more how it would work at the API and schema level?

**Reply by rjmackay** [reply]
Right now we have a schema something like this
```typescript
    Route: {
      pk: { type: String, value: 'route#${routeId}' },
      sk: { type: String, value: 'route' },
      routeId: { type: String, required: true },
      startTime: { type: String, required: true, transform: transformDateTime },
    },
```
where `transformDateTime` handles a Luxon DateTime object and convert to/from a string.

Ideally it might look like
```typescript
    Route: {
      pk: { type: String, value: 'route#${routeId}' },
      sk: { type: String, value: 'route' },
      routeId: { type: String, required: true },
      startTime: { type: DateTime, required: true },
    },
```
then have a `customTypes` option to the Table ie.
```typescript
{
  customTypes: {
      'DateTime': {
        type: String, // The underlying storage 
        transformWrite: (val) => DateTime.fromISO(val),  // Called to transform DateTime to String
        transformRead: (val) => val.toISO(),  // Called to transform String to DateTime
      }
  }
}
```

As an example, the built in Date handling might look something like this:
```typescript
{
  customTypes: {
      'Date': {
        type: String, // The underlying storage 
        transformWrite: (val) => { return val.toISOString() },  // Called to transform DateTime to String
        transformRead: (val) => { return new Date(val) },  // Called to transform String to DateTime
      }
  }
}
```

**Comment by rjmackay** [comment]
Does this also remove the ability to use a function as the `value` for a field?
I have a bunch of GSIs where I use a custom value function to switch the index structure depending on some of the model properties. ie. we return null if the models `deletedAt` prop is set, and index differently if the model is assigned an owner or not.

Even if I have to simplify these, what's the best way to implement a sparse index?

**Reply by mobsense** [reply]
Yes. 

That was in the code previously, but undocumented and is now flagged as DEPRECATED. 

You can use a transform function in the API call via params or at the table level. Via either, you can change the SK based on the properties. The table transform works on the entire item. The API level transform is called for each field in turn, but you have access to the properties via the model.properties parameter.

Not quite sure how that relates to sparse indexes. Here are some thoughts about the topic: 

https://www.sensedeep.com/blog/posts/2021/sparse-gsi-indexes.html

**Reply by rjmackay** [reply]
Here's an example of what we're doing now:
```typescript
      gs1pk: {
        type: String,
        value: (_propertyName: string, properties: Record<string, unknown>): string => {
          if (properties.deletedAt) {
            // If route is deleted, remove from index
            return null
          }
          // Overload the GS1 index, since we know there's never overlap between claimed/unclaimed
          else if (properties.claimedDriverId) {
            // Use driverId if claimed
            return `driver#${properties.claimedDriverId}`
          } else {
            // Use pk if unclaimed
            return `zone#${properties.launchZoneId}#${properties.deliveryDate}`
          }
        },
      },
```

This is pretty model specific, can I use a transform function at the model level? or only table level?

**Reply by mobsense** [reply]
Thanks for the example, pretty cool use of GSIs.

There are 2 ways to do this and we're thinking deeply about what is the best way to centralize this kind of code, without making Schemas contain code.

We're thinking about how to best accommodate this and still keep Schemas literal without code.

There are currently 2 options. Unfortunately blocking issues prevent them being used yet, but I'd like to get your feedback.

1.  Provide a function for the property. This currently works for Javascript and Typescript if you use (as any).

We could possibly change the type system to remove the 'as any' requirement. There may be a way currently (still researching this) to put the 'as any' on the specific property.

In your case: name would be gs1pk.

```
    user = await User.create({
        email: 'coy@acme.com',
        domain: 'local',
        name: (): string => {
          return 'coyote'
        }
    } as any)
```

The downside of this is you don't have access to the properties and need to access them from your JS scope chain.

2. The second way is via a transform. Unfortunately this is currently not in the TS definitions, but is in the Javascript code.

```
    user = await User.create({
        email: 'coy@acme.com',
        domain: 'local',
    }, {
      transform(model: UserType, op: string, name: string, value: any): string {
        return 'road-runner'
      }
    })
```

Similar downside that you don't have direct access to other properties expressed.

The other downside of these approaches is that the code may need to be replicated if you have code scattered that needs the same logic.

**Comment by mobsense** [comment]
Here is a proposal:

We make the value template function be provided in params using the same calling sequence.

```
    user = await User.create({
        ....
    }, {
        value(name: string, properties: OneProperties): string {
          if (properties.deletedAt) {
            // If route is deleted, remove from index
            return null
          }
          // Overload the GS1 index, since we know there's never overlap between claimed/unclaimed
          else if (properties.claimedDriverId) {
            // Use driverId if claimed
            return `driver#${properties.claimedDriverId}`
          } else {
            // Use pk if unclaimed
            return `zone#${properties.launchZoneId}#${properties.deliveryDate}`
          }
        },
})
```

**Reply by mobsense** [reply]
Having same discussion on #162.

Can we unify in #162?
